File: majorityElement.py
# problem: https: // leetcode.com/problems/majority-element/submissions/
# Runtime: 176 ms, faster than 19.30 % of Python3 online submissions for Majority Element.
# Memory Usage: 15.6 MB, less than 18.99 % of Python3 online submissions for Majority Element.

from typing import List


class Solution:
    def majorityElement(self, nums: List[int]) -> int:

        # from statistics import mode
        # return mode(nums)
        from collections import defaultdict
        count = defaultdict(int)
        mode = (len(nums) // 2) + 1
        for num in nums:
            count[num] += 1
            if count[num] >= mode:
                return num

File: majorityElement.go
// problem: https://leetcode.com/problems/majority-element/
// Runtime: 13 ms, faster than 92.51% of Go online submissions for Majority Element.
// Memory Usage: 6.3 MB, less than 48.46% of Go online submissions for Majority Element.

package algorithms

func majorityElement(nums []int) int {
	sol, count := 0, 0
	for _, num := range nums {
		if count == 0 {
			sol = num
		}
		if num == sol {
			count++
		} else {
			count--
		}
	}
	return sol
}
